app/lib/history/models.dart

```diff
@@ -122,46 +128,72 @@ abstract class HistoryData {
   DateTime get timestamp;
 }
 
+// ignore: one_member_abstracts
 abstract class HistoryEvent {
-  String getType();
-
-  Map<String, dynamic> toJson();
-
   String formatMarkdown(HistoryData data);
 }
 
-typedef HistoryEvent HistoryEventFromJson(Map<String, dynamic> json);
+@JsonSerializable()
+class HistoryUnion {
```

Comment:
What if we dropped the abstract class `HistoryEvent` and called this `HistoryEvent`?